### ISSUE TITLE
ArrayIndexOutOfBoundsException  in TopN query

### ISSUE THREAD

**Issue by himanshug** [apache/incubator-druid#4937]
I'm seeing following in KafkaIndexTask peons time to time, looks like a bug.

```
2017-10-10T18:19:35,062 ERROR [topN_xx_metrics_cluster_[2017-10-10T00:00:00.000Z/2017-10-10T18:20:00.000Z]] io.druid.query.ChainedExecutionQueryRunner - Exception with one of the sequences!
java.lang.ArrayIndexOutOfBoundsException: 67
        at io.druid.query.topn.Generic1AggPooledTopNScannerPrototype$Copy0.scanAndAggregate(Generic1AggPooledTopNScannerPrototype.java:48) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.PooledTopNAlgorithm.scanAndAggregateGeneric1Agg(PooledTopNAlgorithm.java:372) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.PooledTopNAlgorithm.lambda$computeSpecializedScanAndAggregateImplementations$2(PooledTopNAlgorithm.java:171) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.PooledTopNAlgorithm.scanAndAggregate(PooledTopNAlgorithm.java:319) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.PooledTopNAlgorithm.scanAndAggregate(PooledTopNAlgorithm.java:54) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.BaseTopNAlgorithm.runWithCardinalityKnown(BaseTopNAlgorithm.java:118) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.BaseTopNAlgorithm.run(BaseTopNAlgorithm.java:81) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.TopNMapFn.apply(TopNMapFn.java:105) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.TopNQueryEngine$1.apply(TopNQueryEngine.java:100) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.topn.TopNQueryEngine$1.apply(TopNQueryEngine.java:93) ~[druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.MappingAccumulator.accumulate(MappingAccumulator.java:40) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.MappingAccumulator.accumulate(MappingAccumulator.java:40) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.BaseSequence.accumulate(BaseSequence.java:46) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.FilteredSequence.accumulate(FilteredSequence.java:45) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.java.util.common.guava.Sequences.toList(Sequences.java:150) ~[java-util-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.ChainedExecutionQueryRunner$1$1$1.call(ChainedExecutionQueryRunner.java:130) [druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at io.druid.query.ChainedExecutionQueryRunner$1$1$1.call(ChainedExecutionQueryRunner.java:120) [druid-processing-0.11.1-1507587363-0b733f3-76.jar:0.11.1-1507587363-0b733f3-76]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_131]
```

**Comment by leventov** [comment]
Seems to me that this is because column values are added while query is in progress. Related to #4049.

**Comment by gianm** [comment]
Recently I also saw this maxId-checking ISE in StringDimensionIndexer get thrown on a realtime task:

```
      @Override
      public String lookupName(int id)
      {
        if (id >= maxId) {
          throw new ISE("id[%d] >= maxId[%d]", id, maxId);
        }
        final String strValue = getActualValue(id, false);
        return extractionFn == null ? strValue : extractionFn.apply(strValue);
      }
```

I (unfortunately) lost the stack trace, but it would be also explained by some bug related to #4049.

**Comment by himanshug** [comment]
#4049 is included in 0.11.0 release, so marking this bug for 0.11.0 milestone.

**Comment by himanshug** [comment]
@kaijianding can you please take a look?

**Comment by gianm** [comment]
@himanshug #4049 was included in 0.10.1, so it's not a new bug in this release; imo that means we should be willing to ship 0.11.0 without the fix if we cannot get it fixed quickly.

**Comment by himanshug** [comment]
i see, didn't check the 0.10.1 branch and looked only in 0.11.0.
ok, we may not want to block 0.11.0 release. but, we should not let this slide away, so marking it for 0.11.1 . however, it would be nice to have it fixed in 0.11.0 if possible.

**Comment by gianm** [comment]
Fixed by #6340 (hopefully).
